File: Arcade/Intro/C++/60_sudoku.cpp
bool sudoku(std::vector<std::vector<int>> grid) {
    map<int, int> colCount;
    for(int c=0;c<grid[0].size();c++) {
        colCount=map<int, int>();
        for(int r=0;r<grid.size();r++) {
            colCount[grid[r][c]]++;
            if(colCount[grid[r][c]]>1)   return false;
        }
    }
    map<int, int> rowCount;
    for(int r=0;r<grid.size();r++) {
        rowCount=map<int, int>();
        for(int c=0;c<grid[0].size();c++) {
            rowCount[grid[r][c]]++;
            if(rowCount[grid[r][c]]>1)    return false;
        }
    }
    for(int r=0;r<grid.size();r+=3) {
        for(int c=0;c<grid[0].size();c+=3) {
            map<int, int> subCount;
            for(int i=0;i<3;i++) {
                for(int j=0;j<3;j++) {
                    subCount[grid[r+i][c+j]]++;
                    if(subCount[grid[r+i][c+j]]>1) return false;
                }
            }
        }
    }
    return true;
}

File: Arcade/Intro/JavaScript/60_sudoku.js
function sudoku(grid) {
    var columns = [];
    for (var i = 0; i < grid.length; i++) {
        columns.push([grid[0][i],grid[1][i],grid[2][i],grid[3][i],grid[4][i],
                     grid[5][i],grid[6][i],grid[7][i],grid[8][i]]);
    }
    for (var i = 0; i < grid.length; i++) {
        if (i % 3 === 0) {
            for (var j = 0; j < grid.length/3; j++) {
                var col = j*3;
                var square = [grid[i][col],grid[i][col+1],grid[i][col+2],
                              grid[i+1][col],grid[i+1][col+1],grid[i+1][col+2],
                              grid[i+2][col],grid[i+2][col+1],grid[i+2][col+2]];
                if (unique(square) === false) {
                    return false;
                }
            }
        }
        if (unique(grid[i]) === false) {
            return false;
        } else if (unique(columns[i]) === false ) {
            return false;
        }
    }
    return true;
}

function unique(array) {
    array.sort();
    for (i = 0; i < array.length; i++) {
        if (array[i] !== i + 1) {
            return false;
        }
    }
    return true;
}
